File: skellig-test-step-processing/src/main/java/org/skellig/teststep/processing/runner/TestStepsRegistry.java
package org.skellig.teststep.processing.runner;

import org.skellig.teststep.processing.exception.TestStepRegistryException;
import org.skellig.teststep.processing.model.TestStepFileExtension;
import org.skellig.teststep.reader.TestStepReader;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

class TestStepsRegistry {

    private Collection<Map<String, Object>> testSteps;
    private TestStepFileExtension testStepFileExtension;
    private Map<String, Pattern> stepNamePatternsCache;
    private TestStepReader testStepReader;
    private Collection<Path> testStepsPaths;

    TestStepsRegistry(TestStepFileExtension testStepFileExtension,
                      TestStepReader testStepReader) {
        this.testStepFileExtension = testStepFileExtension;
        this.testStepReader = testStepReader;
        stepNamePatternsCache = new ConcurrentHashMap<>();
    }

    void registerFoundTestStepsInPath(Collection<Path> testStepsPaths) {
        this.testStepsPaths = testStepsPaths;
        testSteps = getTestStepsFromPath(testStepsPaths);
    }

    Optional<Map<String, Object>> getByName(String testStepName) {
        return testSteps.parallelStream()
                .filter(testStep -> getPatternOfTestStep(getTestStepName(testStep)).matcher(testStepName).matches())
                .findFirst();
    }

    Map<String, String> extractParametersFromTestStepName(Map<String, Object> rawTestStep, String testStepName) {
        Map<String, String> parameters = new HashMap<>();
        Matcher matcher = getPatternOfTestStep(getTestStepName(rawTestStep)).matcher(testStepName);
        if (matcher.find()) {
            for (int i = 1; i <= matcher.groupCount(); i++) {
                parameters.put("$" + i, matcher.group(i));
            }
        }
        return parameters;
    }

    private String getTestStepName(Map<String, Object> rawTestStep) {
        return String.valueOf(rawTestStep.get("name"));
    }

    Collection<Path> getTestStepsRootPath() {
        return testStepsPaths;
    }

    private Pattern getPatternOfTestStep(String testStepName) {
        return stepNamePatternsCache.computeIfAbsent(testStepName, v -> Pattern.compile(testStepName));
    }

    private Collection<Map<String, Object>> getTestStepsFromPath(Collection<Path> rootPaths) {
        return rootPaths.stream()
                .map(rootPath -> {
                    try {
                        return Files.walk(rootPath)
                                .parallel()
                                .filter(path -> String.valueOf(path.getFileName()).endsWith(testStepFileExtension.getName()))
                                .map(filePath -> testStepReader.read(filePath))
                                .flatMap(Collection::stream)
                                .collect(Collectors.toList());
                    } catch (IOException e) {
                        throw new TestStepRegistryException(e.getMessage(), e);
                    }
                })
                .flatMap(Collection::stream)
                .collect(Collectors.toList());
    }
}

File: skellig-test-step-runner/src/main/kotlin/org/skellig/teststep/runner/TestStepsRegistry.kt
package org.skellig.teststep.runner

import org.skellig.teststep.processing.model.factory.TestStepRegistry
import org.skellig.teststep.processing.util.CachedPattern.Companion.compile
import org.skellig.teststep.reader.TestStepReader
import org.skellig.teststep.runner.exception.TestStepRegistryException
import org.skellig.teststep.runner.model.TestStepFileExtension
import java.net.URI
import java.nio.file.*
import java.nio.file.attribute.BasicFileAttributes
import java.util.stream.Collectors

internal class TestStepsRegistry(private val testStepFileExtension: TestStepFileExtension,
                                 private val testStepReader: TestStepReader) : TestStepRegistry {

    private var testSteps: Collection<Map<String, Any?>> = emptyList()

    fun registerFoundTestStepsInPath(testStepsPaths: Collection<URI>) {
        testSteps = getTestStepsFromPath(testStepsPaths)
    }

    override fun getByName(testStepName: String): Map<String, Any?>? =
        testSteps.parallelStream()
            .filter { testStep: Map<String, Any?> ->
                compile(getTestStepName(testStep)).matcher(testStepName).matches()
            }
            .findFirst()
            .orElse(null)

    override fun getTestSteps(): Collection<Map<String, Any?>> = testSteps

    private fun getTestStepName(rawTestStep: Map<String, Any?>): String =
        rawTestStep["name"]?.toString() ?: error("Attribute 'name' was not found in a raw Test Step $rawTestStep")

    private fun getTestStepsFromPath(rootPaths: Collection<URI>): Collection<Map<String, Any?>> {
        val readFileStrategy = RawTestStepsReaderStrategy()
        return rootPaths
            .map { readFileStrategy.getTestStepsFromUri(it) }
            .flatten()
            .toList()
    }

    private inner class RawTestStepsReaderStrategy : RawTestStepsReader {
        private val fileReader = RawTestStepsFromFileReader()
        private val jarReader = RawTestStepsFromJarFileReader()

        override fun getTestStepsFromUri(rootUri: URI): Collection<Map<String, Any?>> =
            try {
                if (rootUri.toURL().protocol == "jar") jarReader.getTestStepsFromUri(rootUri)
                else fileReader.getTestStepsFromUri(rootUri)
            } catch (ex: Exception) {
                throw TestStepRegistryException(ex.message, ex)
            }
    }

    private interface RawTestStepsReader {
        fun getTestStepsFromUri(rootUri: URI): Collection<Map<String, Any?>>
    }

    private open inner class RawTestStepsFromFileReader : RawTestStepsReader {

        override fun getTestStepsFromUri(rootUri: URI): Collection<Map<String, Any?>> =
            walkThroughFiles(Paths.get(rootUri))

        protected fun walkThroughFiles(root: Path): Collection<Map<String, Any?>> =
            Files.walk(root)
                .parallel()
                .filter { it.toString().endsWith(testStepFileExtension.extension) }
                .map { testStepReader.read(it.toUri().toURL().openStream()) }
                .flatMap { it.stream() }
                .collect(Collectors.toList())
    }

    private inner class RawTestStepsFromJarFileReader : RawTestStepsFromFileReader() {

        override fun getTestStepsFromUri(rootUri: URI): Collection<Map<String, Any?>> {
            val rawTestSteps = mutableListOf<Map<String, Any?>>()
            FileSystems.newFileSystem(rootUri, emptyMap<String, Any>()).use {
                for (root in it.rootDirectories) {
                    rawTestSteps.addAll(walkThroughFiles(root))
                }
            }
            return rawTestSteps
        }
    }
}